src/asl/asl-graph.ts

```diff
@@ -2037,13 +2046,102 @@ export namespace ASLGraph {
    *
    * https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-intrinsic-functions.html#amazon-states-language-intrinsic-functions-escapes
    */
-  export function escapeFormatString(literal: ASLGraph.LiteralValue) {
-    if (typeof literal.value === "string") {
-      return literal.value.replace(/[\}\{\'}]/g, "\\$&");
+  export function escapeFormatLiteral(literal: ASLGraph.LiteralValue) {
+    if (ASLGraph.isLiteralString(literal)) {
+      return escapeFormatLiteralString(literal);
     } else {
       return literal.value;
     }
   }
+
+  /**
+   * Turns any literal value in a string or intrinsic functions that can generate a string representation.
+   *
+   * ```ts
+   * literalValue("value") // => "value"
+   * literalValue(1) // => "1"
+   * literalValue({ a: 1 }) // => "{ \"a\": "1" }""
+   * literalValue({ "a.$": "$.var" }) => intrinsicFormat('{ "a": {} }', JsonToString("$.var"))
+   * literalValue(["$.var"]) => intrinsicFormat('[{}]', JsonToString("$.var"))
+   * ```
+   */
+  export function stringifyLiteral(
+    value: ASLGraph.LiteralValue
+  ): ASLGraph.IntrinsicFunction | ASLGraph.LiteralValue<string> {
+    if (ASLGraph.isLiteralNull(value)) {
+      return ASLGraph.literalValue("null");
+    } else if (ASLGraph.isLiteralArray(value)) {
+      return stringifyLiteralArray(value);
+    } else if (ASLGraph.isLiteralObject(value)) {
+      return stringifyLiteralObject(value);
+    }
+    return ASLGraph.isLiteralString(value)
+      ? value
+      : ASLGraph.literalValue(String(value.value));
+
+    function stringifyLiteralObject(
+      value: ASLGraph.LiteralValue<Record<string, ASLGraph.LiteralValueType>>
+    ): ASLGraph.IntrinsicFunction | ASLGraph.LiteralValue<string> {
+      if (!value.containsJsonPath) {
+        return ASLGraph.literalValue(JSON.stringify(value.value), false);
+      }
+      const entries = Object.entries(value.value).map(([key, value]) => {
+        if (key.endsWith(".$")) {
+          // assuming the these values are
```

Comment:
Are what?

Reply:
sigh... half thoughts.

assuming that the keys will contain json path and not intrinsic functions.

I am starting to think of a re-write where literal values contain the JsonPath and IntrinsicFunction structs instead of strings and then resolving at the end. That way we don't need the "containsJsonPath", basically leave all resolution to the end.

Also thinking of introducing intrinsic function as an Output.  This would allow for optimal programs because intrinsic functions take 2 states to resolve today, but negative states when they are composed.

for example as array returns an intrinsic function

```ts
{
    v: [input.value]
}
```

```ts
// format the array
{
    Parameters: { "arr.$": "States.Array($.value)" },
    ResultPath: "$.t",
    output: { jsonPath: "$.t.arr" }
}
```

```ts
// create the object with array prop
{
    Parameters: { "v.$": "$.t.arr" }
}
```

But if Array returned an intrinsic function instead...

```ts
{
    output: { funcName: "States.Array", args: [{ josnPath: "$.value" }] }
}
```

```ts
{
    Parameters: { "v.$": "States.Array($.value)" }
}
```